torch/jit/__init__.py

```diff
@@ -669,6 +534,119 @@ def script_method(fn):
     return ScriptMethodStub(createResolutionCallback(), get_jit_ast(fn))
 
 
+# These OrderedDictWrapper classes replace the actual OrderedDicts in
+# module with versions that get/set properties inside of script::Module.
+# This allows us to reuse most of nn.Module while still storing the
+# data in C++.
+# Each OrderedDict needs to support:
+#  x not in view
+#  x in view
+#  view[name] = ...
+#  view.values()
+#  del view[name]
+#  view.items()
+#  view.keys()
+
+class OrderedDictWrapper(object):
+    def __init__(self, module):
+        self.module_ref = weakref.ref(module)
+
+    @property
+    def module(self):
+        r = self.module_ref()
+        if r is None:
+            raise RuntimeError("_parameters or _modules alive after module is dead")
+        return r
+
+    def keys(self):
+        return [k for k, v in self.items()]
+
+    def values(self):
+        return [v for k, v in self.items()]
+
+    def __delitem__(self, k):
+        raise RuntimeError("cannot delete methods or parameters of a script module")
+
+    def items(self):
+        raise NotImplementedError
+
+    def __contains__(self, k):
+        raise NotImplementedError
+
+    def __getitem__(self, k):
+        raise NotImplementedError
+
+    def __setitem__(self, k, v):
+        raise NotImplementedError
+
+
+class OrderedModuleDict(OrderedDictWrapper):
+    def __init__(self, module):
+        super(OrderedModuleDict, self).__init__(module)
+
+    def items(self):
+        r = self.module._get_modules()
+        return r
+
+    def __contains__(self, k):
+        return self.module._has_module(k)
+
+    # TODO: allow for python modules in addition to script modules
+    def __setitem__(self, k, v):
+        self.module._register_module(k, v)
+        # note: script modules are subclassed in python and the
+        # C++ script::Module class will not hold references to them
+        # to ensure that you always get the same python value here
+        # we store it as a native attribute _in addition to_
+        # registering it with the C++ script::Module
+        object.__setattr__(self.module, k, v)
```

Comment:
It looks like this is unnecessary or broken. We never access attributes of this dict object like we did when this was in module